addons/account/models/account_payment.py

```diff
@@ -4,15 +4,74 @@
+    _name = "account.payment.method.type"
+    _description = "Method Types"
+
+    name = fields.Char(translate=True, required=True)
+    code = fields.Char(required=True)  # For internal identification
+    category = fields.Selection(selection=[('electronic', 'Electronic'), ('manual', 'Manual')],
+                                required=True)
+    payment_type = fields.Selection(selection=[('inbound', 'Inbound'), ('outbound', 'Outbound')],
+                                    required=True)
+    journal_type = fields.Selection(selection=[('bank', 'Bank'), ('cash', 'Cash')],
+                                    help='Choose on which journal type this method will be usable. '
+                                         'Leave empty to allow both types.',
+                                    required=False)
+
+    _sql_constraints = [
+        ('name_code_unique', 'unique (code, payment_type)', 'The combination code/payment type already exists!'),
+    ]
+
+
 class AccountPaymentMethod(models.Model):
     _name = "account.payment.method"
     _description = "Payment Methods"
-    _order = 'sequence'
 
-    name = fields.Char(required=True, translate=True)
-    code = fields.Char(required=True)  # For internal identification
-    payment_type = fields.Selection([('inbound', 'Inbound'), ('outbound', 'Outbound')], required=True)
-    sequence = fields.Integer(help='Used to order Methods in the form view', default=10)
+    name = fields.Char(compute='_compute_name', readonly=False, store=True)
+    active = fields.Boolean(default=True)
+    sequence = fields.Integer()
+    category = fields.Selection(
+        selection=[('electronic', 'Electronic'), ('manual', 'Manual')],
+        compute='_compute_category', readonly=False,
+        required=True, search='_search_category'
+    )
+    method_type = fields.Many2one(
+        string='Payment Method',
+        comodel_name='account.payment.method.type',
+        domain="[('payment_type', '=?', payment_type), ('category', '=?', category), ('id', 'in', available_method_type_ids)]",
+        required=True
+    )
+    payment_account_id = fields.Many2one(
+        comodel_name='account.account',
+        check_company=True,
+        copy=False,
+        ondelete='restrict',
+        domain=lambda self: "[('deprecated', '=', False), ('company_id', '=', company_id), \
+                             ('user_type_id.type', 'not in', ('receivable', 'payable')), \
+                             '|', ('user_type_id', '=', %s)]" % self.env.ref('account.data_account_type_current_assets').id
+
+    )
+    code = fields.Char(related='method_type.code', readonly=True)
+    payment_type = fields.Selection(related='method_type.payment_type', readonly=True)
+    journal_id = fields.Many2one(comodel_name='account.journal', required=True)
+    company_id = fields.Many2one(related='journal_id.company_id')
+    available_method_type_ids = fields.Many2many(related='journal_id.available_method_type_ids')
```

Comment:
not sure about all these related fields. Btw you should split the fields like `Business fields` / `Display purpose fields`